algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_jax/models.py

```diff
@@ -0,0 +1,730 @@
+        lambda x: x[jnp.arange(inputs.shape[0]), lengths - 1], cell_states)
+
+    if reverse:
+      outputs = flip_sequences(outputs, lengths)
+
+    return outputs, final_state
+
+
+class GenericRNN(nn.Module):
+  """Generic RNN class.
+
+  This provides generic RNN functionality to encode sequences with any RNN cell.
+  The class provides unidirectional and bidirectional layers, and these are
+  stacked when asking for multiple layers.
+
+  This class be used to create a specific RNN class such as LSTM or GRU.
+
+  Attributes:
+    cell_type: An RNN cell class to use, e.g., `flax.linen.LSTMCell`.
+    hidden_size: The size of each recurrent cell.
+    num_layers: The number of stacked recurrent layers. The output of the first
+      layer, with optional dropout applied, feeds into the next layer.
+    dropout_rate: Dropout rate to be applied between LSTM layers. Only applies
+      when num_layers > 1.
+    recurrent_dropout_rate: Dropout rate to be applied on the hidden state at
+      each time step repeating the same dropout mask.
+    bidirectional: Process the sequence left-to-right and right-to-left and
+      concatenate the outputs from the two directions.
+    cell_kwargs: Optional keyword arguments to instantiate the cell with.
+  """
+  cell_type: Type[nn.recurrent.RNNCellBase]
+  hidden_size: int
+  num_layers: int = 1
+  dropout_rate: float = 0.
+  recurrent_dropout_rate: float = 0.
+  bidirectional: bool = False
+  cell_kwargs: Mapping[str, Any] = flax.core.FrozenDict()
+
+  @nn.compact
+  def __call__(
+      self,
+      inputs: Array,
+      lengths: Array,
+      initial_states: Optional[Sequence[StateType]] = None,
+      deterministic: bool = False) -> Tuple[Array, Sequence[StateType]]:
+    """Processes the input sequence using the recurrent cell.
+
+    Args:
+      inputs: The input sequence <float32>[batch_size, sequence_length, ...]
+      lengths: The lengths of each sequence in the batch. <int64>[batch_size]
+      initial_states: The initial states for the cells. You must provide
+        `num_layers` initial states (when using bidirectional, `num_layers *
+        2`).
+        These must be ordered in the following way: (layer_0_forward,
+          layer_0_backward, layer_1_forward, layer_1_backward, ...). If None,
+          all initial states will be initialized with zeros.
+      deterministic: Disables dropout between layers when set to True.
+    Returns:
+      The sequence of all outputs for the final layer, and a list of final
+      states for each cell and direction. Directions are alternated (first
+      forward, then backward, if bidirectional). For example for a bidirectional
+      cell this would be: layer 1 forward, layer 1 backward, layer 2 forward,
+      layer 2 backward, etc..
+      For some cells like LSTMCell a state consists of an (c, h) tuple, while
+      for others cells it only contains a single vector (h,).
+    """
+    batch_size = inputs.shape[0]
+    final_states = []
```

Comment:
we never use `final_states`, so we should be able to delete this

Reply:
deleted final_states